File: Greedy Algorithms, Minimum Spanning Trees, and Dynamic Programming/heap.rb
class Heap
    attr_accessor :elements

    def initialize
        @elements=[]
        @positions=Hash.new
    end

    def findMin
        @elements[0]
    end

    def extractMin
        self.delete(@elements[0][1])
    end

    def heapify(arr)
        @elements = arr
        @positions = Hash[@elements.collect.with_index {|e, i| [e[1], i]}]
        (1..arr.size).each do |i|
            self.bubbleDown(i)
        end
    end

    def insert(element)
        @elements.push(element)
        @positions[element[1]] = (@elements.size) - 1 
        self.bubbleUp
        element
    end

    def delete(element)
        position = @positions[element]
        @elements[position], @elements[-1], @positions[element], @positions[@elements[-1][1]] = @elements[-1], @elements[position] , @positions[@elements[-1][1]], @positions[element]
        deleted = @elements.pop
        @positions.delete(element)
        self.bubbleDown(position+1)
        deleted
    end

    def to_s
        "elements: #{@elements.map{ |pair| "heap key: #{pair[0]} - position #{@positions[pair[1]]} - element: #{pair[1].to_s}"}.to_s}"
    end

    def empty?
        @elements.empty?
    end

    def include?(element)
        @positions.has_key?(element)
    end

    private

    def bubbleUp
        position = @elements.size
        parentPosition = getParentPosition(position)
        while @elements[parentPosition][0] > @elements[position-1][0]
            @elements[parentPosition], @elements[position-1] , @positions[@elements[parentPosition][1]], @positions[@elements[position-1][1]] = @elements[position-1], @elements[parentPosition], @positions[@elements[position -1][1]], @positions[@elements[parentPosition][1]]
            position = parentPosition+1
            return if parentPosition == 0
            parentPosition = getParentPosition(position)
        end
    end

    def bubbleDown(startPosition)
        position = startPosition
        smallerChildPosition = getSmallerChildPosition(position)
        return if smallerChildPosition > @elements.size-1
        while @elements[smallerChildPosition][0] < @elements[position-1][0]
            @elements[smallerChildPosition], @elements[position-1] , @positions[@elements[smallerChildPosition][1]], @positions[@elements[position-1][1]] = @elements[position-1], @elements[smallerChildPosition], @positions[@elements[position-1][1]], @positions[@elements[smallerChildPosition][1]]
            position = smallerChildPosition+1
            smallerChildPosition = getSmallerChildPosition(position)
            return if smallerChildPosition > @elements.size-1
        end
    end

    def getParentPosition(position)
        (position/2).floor - 1
    end

    def getLeftChildPosition(position)
        2*position-1
    end

    def getRightChildPosition(position)
        2*position
    end

    def getSmallerChildPosition(position)
        left = getLeftChildPosition(position)
        right = getRightChildPosition(position)

        return left if right > (@elements.size - 1)

        if @elements[left][0] < @elements[right][0]
            left
        else
            right
        end
    end
end

File: Greedy Algorithms, Minimum Spanning Trees, and Dynamic Programming/heap.py
import math


class Heap:
    def __init__(self):
        self.elements = []
        self.positions = {}

    def find_min(self):
        if self.empty():
            return None
        return self.elements[0]

    def get_element(self, element):
        position = self.positions[element]
        return self.elements[position]

    def extract_min(self):
        if self.empty():
            return None
        return self.delete(self.elements[0][1])

    def heapify(self, arr):
        self.elements = arr
        self.positions = {e[1]: i for i, e in enumerate(self.elements)}
        for i in range(len(arr)):
            self.__bubble_down(i)
        for i in range(len(arr)):
            self.__bubble_up(i)

    def insert(self, element):
        self.elements.append(element)
        position = len(self.elements) - 1
        self.positions[element[1]] = position
        self.__bubble_up(position)
        return element

    def delete(self, element):
        position = self.positions[element]

        lastElement = self.elements[-1][1]
        (
            self.elements[position],
            self.elements[-1],
            self.positions[element],
            self.positions[lastElement],
        ) = (
            self.elements[-1],
            self.elements[position],
            self.positions[lastElement],
            self.positions[element],
        )

        deleted = self.elements.pop()
        del self.positions[element]
        self.__solve_conflicts(position)
        return deleted

    def update(self, element):
        position = self.positions[element[1]]
        if element[0] == self.elements[position][0]:
            return
        self.elements[position] = element
        self.__solve_conflicts(position)

    def __repr__(self):
        return f"elements: {[f'heap key: {pair[0]} - position {self.positions[pair[1]]} - element: {pair[1]}' for pair in self.elements]}"

    def __str__(self):
        return self.__repr__()

    def empty(self):
        return len(self.elements) == 0

    def size(self):
        return len(self.elements)

    def include(self, element):
        return element in self.positions

    def __solve_conflicts(self, position):
        parentPosition = self.__get_parent_position(position)

        if position >= len(self.elements):
            return
        if (
            self.__is_root(position)
            or self.elements[parentPosition][0] < self.elements[position][0]
        ):
            self.__bubble_down(position)
        else:
            self.__bubble_up(position)

    def __bubble_up(self, startPosition):
        position = startPosition
        parentPosition = self.__get_parent_position(position)

        while (
            not self.__is_root(position)
            and self.elements[parentPosition][0] > self.elements[position][0]
        ):
            parentElement = self.elements[parentPosition][1]
            childElement = self.elements[position][1]

            (
                self.elements[parentPosition],
                self.elements[position],
                self.positions[parentElement],
                self.positions[childElement],
            ) = (
                self.elements[position],
                self.elements[parentPosition],
                self.positions[childElement],
                self.positions[parentElement],
            )

            position = parentPosition
            parentPosition = self.__get_parent_position(position)

    def __bubble_down(self, startPosition):
        position = startPosition
        smallerChildPosition = self.__get_smaller_child_position(position)
        while (
            smallerChildPosition < len(self.elements)
            and self.elements[smallerChildPosition][0] < self.elements[position][0]
        ):
            smallerChildElement = self.elements[smallerChildPosition][1]
            currentElement = self.elements[position][1]
            (
                self.elements[smallerChildPosition],
                self.elements[position],
                self.positions[smallerChildElement],
                self.positions[currentElement],
            ) = (
                self.elements[position],
                self.elements[smallerChildPosition],
                self.positions[currentElement],
                self.positions[smallerChildElement],
            )
            position = smallerChildPosition
            smallerChildPosition = self.__get_smaller_child_position(position)

    def __get_parent_position(self, position):
        return math.ceil((position / 2.0)) - 1

    def __get_left_child_position(self, position):
        return 2 * position + 1

    def __get_right_child_position(self, position):
        return 2 * position + 2

    def __is_root(self, position):
        return position == 0

    def __get_smaller_child_position(self, position):
        left = self.__get_left_child_position(position)
        right = self.__get_right_child_position(position)

        if right > len(self.elements) - 1:
            return left

        if self.elements[left][0] < self.elements[right][0]:
            return left
        else:
            return right
